File: 프로그래머스/1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.py
def solution(board, moves):
    board_list = [[] for i in range(len(board))]
    stack = []
    result = 0
    
    for i in board[::-1]:
        for j in range(len(i)):
            if i[j] != 0:
                board_list[j].append(i[j])
            
    for index in moves:
        if len(board_list[index - 1]) != 0:
            doll = board_list[index - 1].pop()
        
            if len(stack) != 0 and stack[-1] == doll:
                stack.pop()
                result += 2
            else:
                stack.append(doll)
    
    return result
        

File: 프로그래머스/1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.java
import java.util.ArrayList;
import java.util.Stack;

public class Solution {

    public int solution(int[][] board, int[] moves) {
        ArrayList<Stack<Integer>> boardList = new ArrayList<>();
        Stack<Integer> stack = new Stack<>();
        int result = 0;
        
        for (int i = 0; i < board[0].length; i++) {
            boardList.add(new Stack<>());
        }
        
        for (int i = board.length - 1; i >= 0; i--) {
            for (int j = 0; j < board[i].length; j++) {
                if (board[i][j] != 0) {
                    boardList.get(j).push(board[i][j]);
                }
            }
        }
        
        for (int move:moves) {
            int col = move - 1;
            
            if(!boardList.get(col).isEmpty()) {
                int doll = boardList.get(col).pop();
                
                if (!stack.isEmpty() && stack.peek() == doll) {
                    stack.pop();
                    result += 2;
                } else {
                    stack.push(doll);
                }
            } 
        }
        
        return result;
        
    }
}
